src/read/macinfo.rs

```diff
@@ -0,0 +1,290 @@
+    /// use gimli::{DebugMacInfo, LittleEndian};
+    ///
+    /// # let buf = [1, 0, 95, 95, 83, 84, 68, 67, 95, 95, 32, 49, 0];
+    /// # let read_section_somehow = || &buf;
+    /// let debug_str = DebugMacInfo::new(read_section_somehow(), LittleEndian);
+    /// ```
+    pub fn new(macinfo_section: &'input [u8], endian: Endian) -> Self {
+        Self::from(EndianSlice::new(macinfo_section, endian))
+    }
+}
+
+impl<R: Reader> DebugMacInfo<R> {
+    /// Look up a macro reference the `.debug_macinfo` section by DebugMacinfoOffset.
+    ///
+    /// A macinfo offset points to a list of macro information entries in the `.debug_macinfo` section.
+    /// To handle this, the function returns an iterator.
+    ///
+    /// ```
+    /// use gimli::{DebugMacInfo, DebugMacinfoOffset, LittleEndian};
+    ///
+    /// # let buf = [1, 0, 95, 95, 83, 84, 68, 67, 95, 95, 32, 49, 0];
+    /// # let offset = DebugMacinfoOffset(0);
+    /// # let read_section_somehow = || &buf;
+    /// # let debug_macinfo_offset_somehow = || offset;
+    /// let debug_macinfo = DebugMacInfo::new(read_section_somehow(), LittleEndian);
+    /// for macinfo in debug_macinfo.get_macinfo(debug_macinfo_offset_somehow()) {
+    ///     println!("Found macro info {:?}", macinfo);
+    /// }
+    /// ```
+    pub fn get_macinfo(
+        &self,
+        offset: DebugMacinfoOffset<R::Offset>,
+    ) -> impl Iterator<Item = DebugMacInfoItem<R>> {
+        let mut input = self.section.clone();
+        if let Ok(_) = input.skip(offset.0) {
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: false,
+            }
+        } else {
+            // If the offset is invalid, the iterator is immediately finished.
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: true,
+            }
+        }
+    }
+}
+
+impl<T> DebugMacInfo<T> {
+    /// Create a `DebugMacInfo` section that references the data in `self`.
+    ///
+    /// This is useful when `R` implements `Reader` but `T` does not.
+    ///
+    /// Used by `DwarfSections::borrow`.
+    pub fn borrow<'a, F, R>(&'a self, mut borrow: F) -> DebugMacInfo<R>
+    where
+        F: FnMut(&'a T) -> R,
+    {
+        borrow(&self.section).into()
+    }
+}
+
+impl<R> Section<R> for DebugMacInfo<R> {
+    fn id() -> SectionId {
+        SectionId::DebugMacinfo
+    }
+
+    fn reader(&self) -> &R {
+        &self.section
+    }
+}
+
+impl<R> From<R> for DebugMacInfo<R> {
+    fn from(macinfo_section: R) -> Self {
+        DebugMacInfo {
+            section: macinfo_section,
+        }
+    }
+}
+
+/// an Entry in the `.debug_macinfo` section.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DebugMacInfoItem<R> {
+    /// The end of the unit.
+    EndOfUnit,
+    /// A macro definition.
+    Define {
+        /// The line number where the macro is defined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
```

Comment:
I think we should make it clear that the text is the macro name, arguments, and value.